component/src/main/java/com/celements/photo/container/ImageUrl.java

```diff
@@ -0,0 +1,227 @@
+
+    public @NotNull Builder url(@NotNull String url) throws IllegalImageUrlException {
+      if (Preconditions.checkNotNull(url).startsWith("/")) {
+        this.url = url;
+      } else {
+        throw new IllegalImageUrlException(url);
+      }
+      return this;
+    }
+
+    public @NotNull Builder action(@NotNull String action) {
+      this.action = Preconditions.checkNotNull(action);
+      return this;
+    }
+
+    public @NotNull Builder space(@NotNull String space) {
+      this.space = Preconditions.checkNotNull(space);
+      return this;
+    }
+
+    public @NotNull Builder name(@NotNull String name) {
+      this.name = Preconditions.checkNotNull(name);
+      return this;
+    }
+
+    public @NotNull Builder filename(@NotNull String filename) {
+      this.filename = Preconditions.checkNotNull(filename);
+      return this;
+    }
+
+    public @NotNull Builder query(@NotNull String query) {
+      this.query = Preconditions.checkNotNull(query);
+      return this;
+    }
+
+    public @NotNull Builder width(int width) {
+      this.width = width;
+      return this;
+    }
+
+    public @NotNull Builder height(int height) {
+      this.height = height;
+      return this;
+    }
+
+    public ImageUrl build() {
+      return new ImageUrl(this);
+    }
+
+  }
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ImageUrl.class);
+
+  private final Pattern EXTRACT_URL_PART_PATTERN = Pattern.compile("/([^/?]*)");
+  private final Pattern CELWIDTH_FROM_URL_PATTERN = Pattern.compile("^.*celwidth=(\\d+)(&.*)?$");
+  private final Pattern CELHEIGHT_FROM_URL_PATTERN = Pattern.compile("^.*celheight=(\\d+)(&.*)?$");
+
+  private boolean isParsed;
+  private URL url;
+  private String urlStr;
+  private String action;
+  private String space;
+  private String name;
+  private String filename;
+  private String query;
+  private Integer width;
+  private Integer height;
+
+  private ImageUrl(Builder builder) {
+    this.urlStr = builder.url;
+    this.action = builder.action;
+    this.space = builder.space;
+    this.name = builder.name;
+    this.filename = builder.filename;
+    this.query = builder.query;
+    this.width = builder.width;
+    this.height = builder.height;
+    this.isParsed = false;
+  }
+
+  public String getUrl() {
+    return getContext().getURLFactory().getURL(getUrlInternal(), getContext());
+  }
+
+  public String getExternalUrl() {
+    return getUrlInternal().toString();
+  }
+
+  public @NotNull Optional<String> getAction() {
+    parseUrl();
+    return Optional.fromNullable(action);
+  }
+
+  public @NotNull Optional<String> getSpace() {
+    parseUrl();
+    return Optional.fromNullable(space);
+  }
+
+  public @NotNull Optional<String> getName() {
+    parseUrl();
+    return Optional.fromNullable(name);
+  }
+
+  public @NotNull Optional<String> getFilename() {
+    parseUrl();
+    return Optional.fromNullable(filename);
+  }
+
+  public @NotNull Optional<String> getQuery() {
```

Comment:
add TODO & issue to `getQuery()` that overwritten width/height via builder is not correctly considered here